CERT-8107 Solana Documentation: Interpreting the results (#340)
Certora/Documentation: docs/solana/output.rst

   :name: call-trace-example
      
   .. image:: img/call_trace_detail.png
        :alt: A call trace example
        :align: right
        :width: 300px

      Call trace output showing counterexample values

   Let's analyze this counterexample:

   1. **Nondet Values**: 

        * The prover chose ``101`` for ``amount1`` and implicitly ``2`` for ``amount2``
        * These values are chosen to demonstrate the violation of our assertions
        * ``<?>`` means that the value is not important for the counterexample 

   2. **Variable States**:

        * ``is_great`` is assumed, thus ``1`` (true) as ``amount1`` (101) is indeed greater than 100
        * ``amount1`` is 101, which satisfies our assumption ``cvt_assume!(is_great)``
        * ``is_great2`` is ``0`` (false) because ``amount2`` (2) is not greater than 100
        * ``amount2`` is 2, which satisfies our assumption ``cvt_assume!(amount2 != 10)``

   3. **Assertion Failure**:

        * The assertion ``cvt_assert!(amount1 < 100)`` fails because we have a contradiction:
            * We assumed ``is_great`` is true, which means ``amount1 > 100``
            * But then we assert ``amount1 < 100``

>         * This is impossible to satisfy, hence the counterexample

Reviewer: @phreppo I think we should add a few lines to explain the JTS feature here. Only caveat: we need to resubmit [the job used above](https://prover.certora.com/output/696612/daa2d29059234c459ab0d2c7f6784cde?anonymousKey=cddfa55db76600ea920eea729a454cde61fe278e), @paminacert can you please share the source code with Francesco for him to resubmit it including the uploaded source code?

2. The rule's success is meaningful

Common Sanity Check Results
~~~~~~~~~~~~~~~~~~~~~~~~

1. **Sanity Check PASSED**: Desired outcome - confirms rule isn't vacuously true
2. **Sanity Check WARNING**: Warning sign - indicates contradictory assumptions

See `Sanity <#solana-sanity-section>`_ for more details.

Best Practices
------------

1. Always enable rule sanity checking in your configuration
2. Use ``nice_print!`` macro for clearer counterexample analysis
3. Review full call traces when investigating failures
4. Validate counterexamples against your program's expected state space

Advanced Topics
-------------

Over-approximation Detection
~~~~~~~~~~~~~~~~~~~~~~~~~

When analyzing counterexamples, consider:

1. Initial state feasibility: Is the starting state a reachable state for the code you're analyzing?
2. Transaction sequence validity: Does the computation path make sense, did you overlook a certain scenario?
3. State transition legitimacy: Do the values throughout the computation match? Are the parameters to your rule changing as expected?

If any seem impossible in your actual program, the counterexample might be due to over-approximation.

> The Dump

Reviewer: Generally @urikirsh says we shouldn't talk about TAC dumps to external users. For Solana, this however is the only way to inspect the results. I suggest we use the term TAC dump and not just dump here.

```suggestion
The TAC Dump
```

~~~~~~~~
.. image:: img/dumpl.png
   :alt: Where to find the dump

> The dump output shows the underlying verification conditions in the prover's intermediate representation. It provides detailed insight into how the prover interprets and processes your code.

Reviewer: ```suggestion
The dump output shows the underlying verification conditions in the prover's intermediate representation (TAC). It provides detailed insight into how the prover interprets and processes your code.
```

*****************************

* ``R2 = havoc (0x65)``: Sets ``amount1`` to ``0x65`` (101 in decimal)
* ``assume (R2>=0x65)&&(R2<=2^64 - 1)``: Constrains ``amount1`` to be ≥ 101 and within u64 bounds
* Prints ``amount1`` value via ``CexPrintValues``

Second Variable Setup (Lines 8-10)
******************************

* ``R3 = havoc (0x2)``: Sets ``amount2`` to ``0x2`` (2 in decimal)
* ``assume (R3<=0x9)||((R3>=0xb)&&(R3<=2^64 - 1))``: Implements ``amount2 != 10`` constraint
* Prints ``amount2`` value

Boolean Operations (Lines 11-16)
****************************

* ``R4 = 0x1``: Sets ``is_great`` to true (1)
* ``R6 = R3<=0x64 ? 0x0 : 0x1``: Computes ``is_great2`` based on ``amount2 <= 100``
* Various print commands show the values of variables

Assertion Checking (Lines 17-19)
****************************

* ``B7 = false``: The assertion ``amount1 < 100`` evaluates to false
* Final ``assert false``: Indicates the failure of the verification condition

.. tip::
   Understanding the dump can help you:
     * Debug complex verification conditions
     * Understand how the prover interprets your assumptions
     * Identify where contradictions occur in your rules

.. note::

>     The regular prover view allows to show values in decimals by using the convert to dec option on the upper right corner. 

Reviewer: ```suggestion
    The Rule Report allows to show values in decimals by using the convert to dec option on the upper right corner. 
```
